From 840d6750eca3e691a5d8703df9ed70b8493aa09e Mon Sep 17 00:00:00 2001
From: Bijan Tabatabai <bijan311@gmail.com>
Date: Wed, 27 Oct 2021 15:04:22 -0500
Subject: [PATCH 018/179] Add extra debug info for range overlaps

---
 mm/file_only_mem.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/mm/file_only_mem.c b/mm/file_only_mem.c
index 2222f8633a4f..6ef1f59fe442 100644
--- a/mm/file_only_mem.c
+++ b/mm/file_only_mem.c
@@ -100,6 +100,10 @@ static void insert_new_mapping(struct fom_proc *proc, struct fom_mapping *new_ma
 		if ((new_map->start >= cur->start && new_map->start < cur->end) ||
 			(new_map->end > cur->start && new_map->end <= cur->end)) {
 			pr_err("insert_new_mapping: Attempting to insert overlapping mapping\n");
+			pr_err("insert_new_mapping: old mapping %llx %llx\n",
+				cur->start, cur->end);
+			pr_err("insert_new_mapping: new mapping %llx %llx\n",
+				new_map->start, new_map->end);
 			BUG();
 		}
 
-- 
2.49.0

