From 0a1f052dea3724a461e92f8ff1c19df1573835c4 Mon Sep 17 00:00:00 2001
From: Bijan Tabatabai <bijan311@gmail.com>
Date: Mon, 28 Feb 2022 15:46:19 -0600
Subject: [PATCH 032/179] Add --pmem_size to the setup command

---
 kernel_options | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel_options b/kernel_options
index fc0f43b502dd..0471f6ac910b 100644
--- a/kernel_options
+++ b/kernel_options
@@ -57,4 +57,4 @@ CONFIG_VIRTIO_NET=y
 CONFIG_ISO9660_FS=m
 CONFIG_EXT4_FS=y
 
-"setup_wkspc {MACHINE} {USER} --clone_wkspc --git_user BijanT --host_bmks --host_dep --unstable_device_names --resize_root --secret {TOKEN} --spec_2017 /p/multifacet/infrastructure/spec2017.iso" "setup_kernel {MACHINE} {USER} --branch dev --repo github.com/BijanT/linux-file-only-mem --git_user BijanT --secret {TOKEN} +CONFIG_TRANSPARENT_HUGEPAGE +CONFIG_MM_ECON -CONFIG_PAGE_TABLE_ISOLATION -CONFIG_RETPOLINE +CONFIG_GDB_SCRIPTS +CONFIG_FRAME_POINTERS +CONFIG_IKHEADERS +CONFIG_SLAB_FREELIST_RANDOM +CONFIG_SHUFFLE_PAGE_ALLOCATOR +CONFIG_FS_DAX +CONFIG_DAX +CONFIG_BLK_DEV_RAM +CONFIG_FILE_ONLY_MEM +CONFIG_BLK_DEV_PMEM +CONFIG_ND_BLK +CONFIG_BTT +CONFIG_NVDIMM_PFN +CONFIG_NVDIMM_DAX +CONFIG_X86_PMEM_LEGACY"
+"setup_wkspc {MACHINE} {USER} --clone_wkspc --git_user BijanT --host_bmks --host_dep --unstable_device_names --resize_root --pmem_size 30 --secret {TOKEN} --spec_2017 /p/multifacet/infrastructure/spec2017.iso" "setup_kernel {MACHINE} {USER} --branch dev --repo github.com/BijanT/linux-file-only-mem --git_user BijanT --secret {TOKEN} +CONFIG_TRANSPARENT_HUGEPAGE +CONFIG_MM_ECON -CONFIG_PAGE_TABLE_ISOLATION -CONFIG_RETPOLINE +CONFIG_GDB_SCRIPTS +CONFIG_FRAME_POINTERS +CONFIG_IKHEADERS +CONFIG_SLAB_FREELIST_RANDOM +CONFIG_SHUFFLE_PAGE_ALLOCATOR +CONFIG_FS_DAX +CONFIG_DAX +CONFIG_BLK_DEV_RAM +CONFIG_FILE_ONLY_MEM +CONFIG_BLK_DEV_PMEM +CONFIG_ND_BLK +CONFIG_BTT +CONFIG_NVDIMM_PFN +CONFIG_NVDIMM_DAX +CONFIG_X86_PMEM_LEGACY"
-- 
2.49.0

