From 16004ef033a6f494a8838bbe8a61346fee7a6f6a Mon Sep 17 00:00:00 2001
From: bijan <bijan@localhost>
Date: Tue, 19 Apr 2022 20:53:48 -0600
Subject: [PATCH 040/179] Fixup last commit

---
 arch/x86/lib/usercopy_64.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/x86/lib/usercopy_64.c b/arch/x86/lib/usercopy_64.c
index 426b4368062d..018072f13496 100644
--- a/arch/x86/lib/usercopy_64.c
+++ b/arch/x86/lib/usercopy_64.c
@@ -141,13 +141,13 @@ void memset_flushcache(void *dst, uint8_t data, size_t len) {
 	uint64_t dest = (uint64_t) dst;
 	uint16_t half_word = (data << 8) | data;
 	uint32_t word = (half_word << 16) | half_word;
-	uint64_t dword = (word << 32) | word;
+	uint64_t dword = ((uint64_t)word << 32) | word;
 
 	/* Check for alignment */
 	if (!IS_ALIGNED(dest, 8)) {
 		unsigned l = min_t(unsigned, len, ALIGN(dest, 8) - dest);
 
-		memset(dest, data, l);
+		memset((void *)dest, data, l);
 		clean_cache_range((void *)dest, l);
 		dest += l;
 		len -= l;
@@ -193,6 +193,7 @@ void memset_flushcache(void *dst, uint8_t data, size_t len) {
 		clean_cache_range((void *)dest, len);
 	}
 }
+EXPORT_SYMBOL_GPL(memset_flushcache);
 
 void memcpy_page_flushcache(char *to, struct page *page, size_t offset,
 		size_t len)
-- 
2.49.0

