From 70bc533d22af8bf6f52c52ea18f1343f33d0a7e2 Mon Sep 17 00:00:00 2001
From: bijan <bijan@localhost>
Date: Tue, 11 Oct 2022 13:38:06 -0600
Subject: [PATCH 055/179] Fix error that shows up qhen using GCC 9.4.0

---
 FOMTierFS/fs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/FOMTierFS/fs.c b/FOMTierFS/fs.c
index cbad48d7fec6..b7e9726e1722 100644
--- a/FOMTierFS/fs.c
+++ b/FOMTierFS/fs.c
@@ -379,6 +379,7 @@ static int fomtierfs_populate_dev_info(struct fomtierfs_dev_info *di, struct blo
     long num_pages;
     pfn_t _pfn;
     struct fomtierfs_page *tmp;
+    struct fomtierfs_page *cursor, *temp;
 
     di->bdev = bdev;
     di->daxdev = fs_dax_get_by_bdev(bdev);
@@ -416,7 +417,6 @@ static int fomtierfs_populate_dev_info(struct fomtierfs_dev_info *di, struct blo
 
 err:
     // Free all of the entries we've put in the list so far
-    struct fomtierfs_page *cursor, *temp;
     list_for_each_entry_safe(cursor, temp, &di->free_list, list) {
         list_del(&cursor->list);
         kfree(cursor);
-- 
2.49.0

