From 9bd95d21f780462c75d199bc2020d233d18edf18 Mon Sep 17 00:00:00 2001
From: Bijan Tabatabai <bijan311@gmail.com>
Date: Tue, 25 Oct 2022 12:47:19 -0500
Subject: [PATCH 067/179] Whoops, pushed up a test demotion watermark last
 commit

---
 FOMTierFS/fs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/FOMTierFS/fs.c b/FOMTierFS/fs.c
index d44f2be17ccc..8081043b64d2 100644
--- a/FOMTierFS/fs.c
+++ b/FOMTierFS/fs.c
@@ -835,7 +835,7 @@ static int fomtierfs_fill_super(struct super_block *sb, struct fs_context *fc)
     wake_up_process(sbi->demote_task);
 
     // Make the demotion watermark 2% of the total mem
-    sbi->demotion_watermark = sbi->mem[FAST_MEM].num_pages -10;//* 2 / 100;
+    sbi->demotion_watermark = sbi->mem[FAST_MEM].num_pages * 2 / 100;
     // Make the alloc watermark 1% of the total mem
     sbi->alloc_watermark = sbi->mem[FAST_MEM].num_pages / 100;
     fc->s_fs_info = sbi;
-- 
2.49.0

