From c42139abd98efba96a4a421a3affbf5a16e4b9e8 Mon Sep 17 00:00:00 2001
From: Bijan Tabatabai <bijan311@gmail.com>
Date: Mon, 31 Oct 2022 15:21:14 -0500
Subject: [PATCH 076/179] Get rid of unneeded, constly locks in
 fomtierfs_fallocate

---
 FOMTierFS/fs.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/FOMTierFS/fs.c b/FOMTierFS/fs.c
index a285afb022c9..64d511ccfabd 100644
--- a/FOMTierFS/fs.c
+++ b/FOMTierFS/fs.c
@@ -78,7 +78,6 @@ struct fomtierfs_page *fomtierfs_alloc_page(struct inode *inode, struct fomtierf
     page->last_accessed = true;
     spin_unlock(&page->lock);
 
-
     // Add the page to the active list
     list_add(&page->list, &prim->active_list);
     prim->active_pages++;
@@ -688,11 +687,11 @@ static long fomtierfs_fallocate(struct file *file, int mode, loff_t offset, loff
             return -ENOMEM;
         }
 
-        write_lock(&inode_info->map_lock);
+        // We normally need to grab inode_info->map_lock, but
+        // since this page is being fallocated, it isn't shared yet.
         if (!fomtierfs_insert_page(&inode_info->page_maps, page)) {
             BUG();
         }
-        write_unlock(&inode_info->map_lock);
     }
 
     return 0;
-- 
2.49.0

