From 90f43565b60cf7266e1f3ca4bf2880deefc665bd Mon Sep 17 00:00:00 2001
From: Bijan Tabatabai <bijan311@gmail.com>
Date: Thu, 10 Nov 2022 14:25:04 -0600
Subject: [PATCH 084/179] Clear the accessed bit in promote_one

---
 FOMTierFS/fs.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/FOMTierFS/fs.c b/FOMTierFS/fs.c
index 3ab9db62a911..e1f1df44f058 100644
--- a/FOMTierFS/fs.c
+++ b/FOMTierFS/fs.c
@@ -560,13 +560,13 @@ static void fomtierfs_promote_one(struct fomtierfs_sb_info *sbi, struct fomtierf
     last_accessed = page->last_accessed;
     page->last_accessed = accessed;
 
+    // Reset the accessed bit if we need to
+    if (accessed)
+        fomtierfs_page_mkold(vma, page, virt_addr, pmdp);
+
     // Only promote if the page has been accessed in both of the last
     // couple of checks.
     if (!accessed || !last_accessed) {
-        // The page was not accessed recently, so put it back and move
-        // on to the next one.
-        fomtierfs_page_accessed(page, virt_addr, pmdp);
-
         list_add(&page->list, &slow_dev->active_list);
         slow_dev->active_pages++;
 
-- 
2.49.0

