From ad9470243386a852f53853326fea11358c71a835 Mon Sep 17 00:00:00 2001
From: BIJAN TABATABAI <btabatabai@wisc.edu>
Date: Tue, 31 Oct 2023 13:24:29 -0500
Subject: [PATCH 119/179] Take mmap lock when initializing badger trap

---
 mm/badger_trap.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/mm/badger_trap.c b/mm/badger_trap.c
index 1f8676517704..504722575265 100644
--- a/mm/badger_trap.c
+++ b/mm/badger_trap.c
@@ -180,12 +180,14 @@ void badger_trap_init(struct mm_struct *mm)
 				if((pmd_flags(*pmd) & mask) != mask)
 					continue;
 				address = (i<<PGDIR_SHIFT) + (j<<PUD_SHIFT) + (k<<PMD_SHIFT);
+				mmap_read_lock(mm);
 				vma = find_vma(mm, address);
 				if(vma && pmd_huge(*pmd) && (hugepage_flags_enabled()||hugepage_flags_always()||vma->vm_flags & VM_HUGEPAGE||is_vm_hugetlb_page(vma)))
 				{
 					spin_lock(&mm->page_table_lock);
 					*pmd = pmd_mkreserve(*pmd);
 					spin_unlock(&mm->page_table_lock);
+					mmap_read_unlock(mm);
 					continue;
 				}
 				for(l=0; l<PTRS_PER_PTE; l++)
@@ -203,6 +205,7 @@ void badger_trap_init(struct mm_struct *mm)
 					}
 					user++;
 				}
+				mmap_read_unlock(mm);
 			}
 		}
 	}
-- 
2.49.0

