From 055fcc908e060d21eae6baabafc979fa3e3920ac Mon Sep 17 00:00:00 2001
From: BIJAN TABATABAI <btabatabai@wisc.edu>
Date: Wed, 10 Jan 2024 13:29:54 -0600
Subject: [PATCH 144/179] BandwidthMMFS: Use generic_get_unmapped_area_topdown

---
 BandwidthMMFS/bandwidth.c | 2 +-
 mm/mmap.c                 | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/BandwidthMMFS/bandwidth.c b/BandwidthMMFS/bandwidth.c
index 9550c251936e..296634162701 100644
--- a/BandwidthMMFS/bandwidth.c
+++ b/BandwidthMMFS/bandwidth.c
@@ -158,7 +158,7 @@ const struct file_operations bwmmfs_file_operations = {
     .splice_read = generic_file_splice_read,
     .splice_write = iter_file_splice_write,
     .llseek = generic_file_llseek,
-    .get_unmapped_area = thp_get_unmapped_area,
+    .get_unmapped_area = generic_get_unmapped_area_topdown,
     .fallocate = bwmmfs_fallocate,
 };
 
diff --git a/mm/mmap.c b/mm/mmap.c
index 493e5b36e4f1..eb8ef97e201e 100644
--- a/mm/mmap.c
+++ b/mm/mmap.c
@@ -1825,6 +1825,7 @@ generic_get_unmapped_area_topdown(struct file *filp, unsigned long addr,
 
 	return addr;
 }
+EXPORT_SYMBOL(generic_get_unmapped_area_topdown);
 
 #ifndef HAVE_ARCH_UNMAPPED_AREA_TOPDOWN
 unsigned long
-- 
2.49.0

