From 6d6edbeca45d2a4199b8157a9bb415f57d687d2c Mon Sep 17 00:00:00 2001
From: BIJAN TABATABAI <btabatabai@wisc.edu>
Date: Mon, 15 Jan 2024 10:32:53 -0600
Subject: [PATCH 154/179] Try to fix deadlock thing

---
 ContigMMFS/contig.c | 10 ----------
 1 file changed, 10 deletions(-)

diff --git a/ContigMMFS/contig.c b/ContigMMFS/contig.c
index af021307ca48..75c1a61d918c 100644
--- a/ContigMMFS/contig.c
+++ b/ContigMMFS/contig.c
@@ -220,16 +220,6 @@ static long contigmmfs_fallocate(struct file *file, int mode, loff_t offset, lof
         sbi->num_pages -= folio_nr_pages(region->folio);
         folio_put(region->folio);
 
-        // Clear the range tlb as necessary
-        if (inode_info->mm && inode_info->mm->badger_trap_en && folio_nr_pages(region->folio) >= 8) {
-            spin_lock(&inode_info->mm->range_tlb_lock);
-            tlb_entry = mtree_erase(&inode_info->mm->all_ranges, region->va_start);
-            spin_unlock(&inode_info->mm->range_tlb_lock);
-
-            if (tlb_entry)
-                kfree(tlb_entry);
-        }
-
         next_region = mt_next(&inode_info->mt, region->va_start, ULONG_MAX);
         kfree(region);
 
-- 
2.49.0

